craft_application/services/project.py

```diff
@@ -0,0 +1,452 @@
+        platform: str,
+        build_for: str,
+        build_on: craft_platforms.DebianArchitecture,
+    ) -> None:
+        """Perform expansion of project environment variables.
+
+        :param project_data: The project's yaml data.
+        :param build_for: The architecture to build for.
+        """
+        # We can use "all" directly after resolving:
+        # https://github.com/canonical/craft-parts/issues/1019
+        if build_for == "all":
+            host_arch = craft_platforms.DebianArchitecture.from_host()
+            for target in craft_platforms.DebianArchitecture:
+                if target != host_arch:
+                    build_for = target.value
+                    break
+            else:
+                raise ValueError(
+                    "Could not find an architecture other than the host architecture "
+                    "to set as the build-for architecture. This is a bug in "
+                    f"{self._app.name} or craft-application."
+                )
+            emit.debug(
+                "Expanding environment variables with the architecture "
+                f"{build_for!r} as the build-for architecture because 'all' was "
+                "specified."
+            )
+
+        environment_vars = self._get_project_vars(project_data)
+        partitions = self.get_partitions_for(
+            platform=platform, build_for=build_for, build_on=build_on
+        )
+        work_dir = util.get_work_dir(self._project_dir)
+        project_dirs = craft_parts.ProjectDirs(work_dir=work_dir, partitions=partitions)
+        info = craft_parts.ProjectInfo(
+            application_name=self._app.name,  # not used in environment expansion
+            cache_dir=pathlib.Path(),  # not used in environment expansion
+            arch=build_for,
+            parallel_build_count=util.get_parallel_build_count(self._app.name),
+            project_name=project_data.get("name", ""),
+            project_dirs=project_dirs,
+            project_vars=environment_vars,
+            partitions=partitions,
+        )
+
+        self.update_project_environment(info)
+        craft_parts.expand_environment(project_data, info=info)
+
+    @final
+    def _preprocess(
+        self,
+        *,
+        build_for: str,
+        build_on: str,
+        platform: str,
+    ) -> dict[str, Any]:
+        """Preprocess the project for the given build-on, build-for and platform.
+
+        This method provides a project dict that has gone through any app-specific
+        pre-processing and has had its grammar validated, has not had environment
+        expansion or parts grammar applied.
+
+        This method is for internal use only, such as for getting partitions.
+
+        :param build_for: The target architecture of the build.
+        :param platform: The name of the target platform.
+        :param build_on: The host architecture the build happens on.
+        :returns: A dict containing a the pre-processed project.
```

Comment:
Also can be in a follow up PR.
```suggestion
        :returns: A dict containing the pre-processed project.
```

Reply:
#686